### PR TITLE
fix(integrations): identity missing

### PR DESCRIPTION
**Pull request by scefali**
This fixes a bug caused when a Gitlab (or other integrations) are installed on multiple orgs but with a different user (either a  different Sentry user or a different Gitlab user). The root of the problem that we end up deleting and recreating the `Identity`:
https://github.com/getsentry/sentry/blob/2520046f4ba1f929974672a1f23a9d02bd22d2d0/src/sentry/integrations/pipeline.py#L147-L149

This code path is hit when we try to create an `Identity` object  that generates an `IntegrityError` if either of the two uniqueness constraints is violated:
https://github.com/getsentry/sentry/blob/2520046f4ba1f929974672a1f23a9d02bd22d2d0/src/sentry/models/identity.py#L78

The two uniqueness violations come from two situations:
1. A user who already linked their identity with Gitlab account is attempting another installation with a different Gitlab account (but same integration)
2. A user is attempting a Gitlab installation logged in on a Gitlab account that's already linked to another user

The intention of `reattach` call is to let an existing `Identity` object get updated with a new `external_id` or `user` if that has been changed. We delete old rows that match that `user` or `external_id`:

https://github.com/getsentry/sentry/blob/2520046f4ba1f929974672a1f23a9d02bd22d2d0/src/sentry/models/identity.py#L91-L92

The issue is that any other `OrganizationIntegration` rows for that integration have a `default_auth_id` that points to an `Identity` object that no longer exists.

The solution I have is the following:
1. If we hit an `IntegrityError` when creating the `Identity`, query if we have an `Identity` object for that identity provider (1 to 1 with an integration) and that `external_id`
2. If that exists, throw an error saying that the Gitlab user is already linked  to a different Sentry user
3. If that doesn't exist, update the existing  `Identity`  object with the new `external_id`  and other fields related to that identity (but don't update the user).

The reason why we want to throw an error for the user being linked to a different account is that we don't want the `Identity` object to reference a new user that might not be part of the organization that has the existing installation.

### CODE REVIEW

**Comment by mgaeta** [comment]
I see that `Identity.reattach` is still used in slack/link_identity.py and msteams/link_identity.py. Will those need to be updated to use `update_external_id_and_defaults`?

File: tests/sentry/integrations/test_pipeline.py [hunk 1]
```diff
@@ -225,6 +228,83 @@ def test_existing_identity_becomes_default_auth_on_new_orgintegration(self, *arg
...
+
+        org_integrations = OrganizationIntegration.objects.filter(integration_id=integration.id)
+        identity = Identity.objects.get(idp_id=identity_provider.id, external_id="new_external_id")
+        for org_integration in org_integrations:
+            assert org_integration.default_auth_id == identity.id
+
+    def test_different_user_same_external_id(self, *args):
+        self.provider = GitlabIntegrationProvider
+        new_user = self.create_user()
+        # self.create_member(organization=self.organization, user=new_user, role="admin")
```
Comment (mgaeta) [inline tests/sentry/integrations/test_pipeline.py:275]: Remove extra comment

File: src/sentry/models/identity.py [hunk 2]
```diff
@@ -108,3 +108,19 @@ def reattach(cls, idp, external_id, user, defaults):
             },
         )
         return identity_model
+
+    @classmethod
+    def update_external_id_and_defaults(cls, idp, external_id, user, defaults):
```
Comment (mgaeta) [inline src/sentry/models/identity.py:113]: Can you add a docstring that explains when to use this rather than `reattach`?

**Review by mgaeta — APPROVED** [review]
Looks good 👍

**Comment by scefali** [comment]
> I see that `Identity.reattach` is still used in slack/link_identity.py and msteams/link_identity.py. Will those need to be updated to use `update_external_id_and_defaults`?

@mgaeta Good question. They do not. We don't use a default identity with MSTeams and Slack so we can continue to use `reattach`.

**Pull request merged**